plan/logical_test.go

```diff
@@ -331,9 +331,8 @@ func (MergeFiltersRule) Name() string {
 }
 
 func (MergeFiltersRule) Pattern() plan.Pattern {
-	return plan.Pat(universe.FilterKind,
-		plan.Pat(universe.FilterKind,
-			plan.Any()))
+	return plan.Multi(universe.FilterKind,
+		plan.Single(universe.FilterKind))
```

Comment:
The `Multi(x, Single(y, Single(z, ...)))` pattern is so common here that I wonder if we need users to specify it in their rules. It seems like maybe the planner could take care of how a pattern is evaluated and only allow multiple successors for the root (the most successorly) part of the pattern?

The pattern for a rule describes a contract between the write of the rule and the planner. Wherever they start their pattern, rule writers are saying, "I won't mutate any nodes that succeed this one". And likewise, wherever the pattern terminates, they are saying, "I won't mutate any predecessors to nodes that match these patterns". 

We could add a boolean parameter to `Match` that is `allowMultiSuccessors` and on the initial invocation by the planner it is `true`, and then on the recursive invocations it is always `false`. Seems like this would give us everything we want without having to change any client code or require the rule writer to think about it too much. What do you think? (hope this is clear)

Reply:
Thinking this over a bit more I think the approach in this PR makes sense. Rule writers should be aware of whether their patterns match multiple successors or not. I don't think there's a one-size-fits-all approach that would work.

I'm now wondering about the effect of these changes on rules that eliminate redundant nodes.
```
p = filter
p |> yield("1")
p |> filter(fn: (r) => true) |> yield("2")
```
A rule that removes the redundant filter will need to be succeeded by both yields. I don't think this would happen currently? I think the planner needs to be updated to look at the list of successors of the eliminated node and its predecessor and concatenate them. I think this is probably the error you already saw?

Reply:
https://github.com/influxdata/flux/blob/53c90d5abbffa5c2c91c5e64e262d083ec4dff1a/plan/heuristic_planner.go#L136